src/nvim/decoration_provider.c

```diff
@@ -139,20 +132,18 @@ void decor_providers_invoke_win(win_T *wp, DecorProviders *providers,
                            ? wp->w_botline
                            : MAX(wp->w_topline + wp->w_height_inner, wp->w_botline)));
 
-  for (size_t k = 0; k < kv_size(*providers); k++) {
-    DecorProvider *p = kv_A(*providers, k);
-    if (p && p->redraw_win != LUA_NOREF) {
+  DecorProvider *p;
+  FOREACH_PROVIDER(p, {
+    if (p->redraw_active && p->redraw_win != LUA_NOREF) {
```

Comment:
We might need a second flag, because either the provider was globally disabled in "start". or it was just disabled for the current window (either by on_win or on_line), and should be tried again on the next on_win. Or a single enum with three states, not sure what is simplest.

Reply:
ah yes. Not to forget `active` which currently acts as a global disable.